### ISSUE TITLE
New format for caching latent and Text Encoder outputs

### ISSUE THREAD

**Issue by kohya-ss** [kohya-ss/sd-scripts#1750]
# Overview
- Policy: File management per image, supporting multiple architectures and data types
- Format: Using `.safetensors` format

--- 

# Latent Cache

## Filename Format
```
{original_image_base_name}_{width}x{height}_{architecture}.safetensors
```
Example: `image1_1024x0768_sd.safetensors`

### Tensor Data
The following tensors are stored as needed:
```
latents_{latents_height}x{latents_width}_{data_type}
latents_flipped_{latents_height}x{latents_width}_{data_type}
alpha_mask_{latents_height}x{latents_width}
```

The data type can be one of the following: `fp32`, `bf16`, `fp16`, `float8_e4m3fn`, `float8_e5m2`.

The priority of types is `fp32>bf16=fp16>float8_e4m3fn=fp8_e5m2`, and if a higher-level type has already been stored, a new one will not be stored. fp16 and bf16, and multiple fp8 may coexist.

Also, when saving a higher-level type, the lower-level type is deleted.

`{latents_height}x{latents_width}` is used for multi-resolution training.

### Metadata (JSON format)
```json
{
  "architecture": "sd",  // or "sdxl" etc.
  "width": "1024",
  "height": "768"
  "crop_ltrb_{latents_height}x{latents_width}": "0,10,0,10",  // comma separated values 
  "format_version": "1.0.0"
}
```

## Key Features
1. File Management
   - Individual file management per image
   - Multiple data types consolidated into a single file
   - Quick image size retrieval through filename parsing (`{width}_{height}` part) without opening the file

2. Type Support
   - Support for multiple data types (fp32, fp16, bf16, fp8)
   - Support for multiple fp8 implementations (E4M3, E5M2)

3. Functionality
   - Includes flipped data for flip augmentation
   - Alpha mask retention

---

# Text Encoder Cache

## Filename Format
```
{original_image_base_name}_{architecture}_{max_token_length}{_m}_te.safetensors
```
`_m` indicates that an attention mask has been applied. `max_token_length` is the length for the main Text Encoder (SD3, FLUX for T5XXL).

Example: `image1_sd3_512_te.safetensors`, `image2_sdxl_77_te_safetensors`, `image1_flux_512_m_te.safetensors`

### Tensor Data
Depending on the architecture, the following is an example:
```
lg_out_1_fp16
t5_out_1_fp16
t5_attn_mask_1
lg_out_2_fp16
t5_out_2_fp16
...
```

The key has a suffix at the end that indicates the data type. The type and priority are the same as for latent.

The attention mask is optional. The attention mask is a long tensor, so it doesn't have the data type suffix. 

There is always a suffix indicating the index of the caption before the masked or the data type suffix. The index is 1-based.

The tensor data doesn't have a batch dimension.

### Metadata (JSON format)
```json
{
  "architecture": "sd",  // or "sdxl" etc.
  "caption1": "a caption for the image",
  "caption2": "another caption for the image", // "capiton2", "caption3" ... optional
  "format_version": "1.0.0"
}
```

## Key Features
1. File Management
   - Individual file management per image
   - Multiple data types consolidated into a single file

2. Type Support
   - Support for multiple data types (fp32, fp16, bf16, fp8)
   - Support for multiple fp8 implementations (E4M3, E5M2)

3. Functionality
   - Includes optional attention mask.
   - Supports multiple captions.

--- 

`.safetensors` may be zipped in the future, in which case the extension will be `.zip`.

Please feel free to share your thoughts and suggestions on this specification.

EDIT: Fix fp8 dtype to `float8_e4m3fn` and `float8_e5m2`, `captions` as list. Use `x` instead of `_` between the height and the width to match the existing name. Add `crop_ltrb` to the metadata. Fix the metadata because safetensors can have `Dict[str, str]` only. Change format version to `major.minor.patch`. Add the max token length to the file name for the text encoder cache. The tensor in the text encoder cache doesn't have the batch dimension, because it is cumbersome to read the entire tensor for reading one of the captions, or update part of the tensor. If an attention mask is applied, it is saved to a separate file. Define what model `max_token_length`  refers to.

**Comment by sdbds** [comment]
Recommendations: 
1. Use [ml_dyptes](https://github.com/jax-ml/ml_dtypes) to support different data types.
2. (Optional) Use [lance](https://lancedb.github.io/lance/arrays.html#extension-arrays) as data container storage.

Compared to safetensors, lance has 3 advantages: 
1、version auto-control
2、randomized loads (which can be used for shuffle).
3, for [huggingface datasets support](https://lancedb.github.io/lance/integrations/huggingface.html#lance-huggingface), and [cloud storage directly read imageuri](https://lancedb.github.io/lance/integrations/huggingface.html#lance-huggingface).

More advantages on large datasets.

**Comment by ChristianMayer** [comment]
For the text encoder cache (very important with T5XXL!) it is important that it is working well with multiple captions (already contained in the proposal) and wildcards. It would also be good when it supports shuffling.

The wildcard support is easy when all possible combinations are created in advance and then just treated like multiple captions.
But this would break the current behavior (as I understood it from reading the code) as the probabilities change. Currently each multi-caption-line is chose with same probability and then in this caption the wildcards are randomly selected:
```
photo of foo
image of {bar|baz}
```
would currently lead to

- 50% `photo of foo`
- 25% `image of bar`
- 25% `image of baz`  

and a naive caching instead to

- 33% `photo of foo`
- 33% `image of bar`
- 33% `image of baz`
 
So when this should be a non breaking change you will need to add the probabilities somehow in the file. Or you are making it a breaking change and accept the change of the probabilities. (I'm actually fine with both but it should be documented)

About how to support shuffling I have no idea how to cache that. Except to also precompute (sample) them and put them all in the cache. But then the cache might become very big.

**Comment by ChristianMayer** [comment]
Due to the big text encoders (T5XXL) I think caching them is even more important than it was in the past so that the VRAM can be freed.

Multi captions and wildcards are different things working in the same direction and I think they are supporting each other very well. So having the combination of both is a great thing to have.
I know that currently they are not cached but it would be really great when in future both could be cached (to save VRAM).

Note: It should be possible to preprocess the caption files to convert wildcards to multi-captions (ignoring the impact on probability). But it would be better when the sd-scripts do it themself.

**Comment by ChristianMayer** [comment]
Sure. My point was that the format might need a field for the probability as well. 
Even when the first implementation doesn't need it (most likely initialize it with a uniform distribution)

**Comment by aa956** [comment]
Image latent cache:

1. Is there a need to add other variations apart from `latents_flipped_`?
   Use cases:
   - Brightness / Color variations (monochrome image, boosted saturation, very dark, very light image)
   - Different area crops, possibly rotated, of the same width * height (but this will probably require different captions too and better handled as separate images?)

2. `{original_image_base_name}`: Max 250 characters of basename with illegal characters filtered out? Or known safe string , e.g. `{sha256(original_image_file)}`?
   Use cases:
   - Windows machines where there will be problems with filename lengths > 255.
   - Windows / Linux dual boot (Windows software for image processing and Linux for training): I've once created a filename containing the `|` on NTFS partition mounted on Linux.

Second is optional: printed warning if latent file is not possible to open is good enough.

**Comment by Ice-YY** [comment]
May I ask when this new format and zip function will be implemented in the training script? I am attempting to finetune SD3.5 with a large dataset, and currently, based on the size of the npz files for now, the cache will exceed 10 TB. 😂

**Comment by FurkanGozukara** [comment]
if we are training with different captions on same dataset, we should use different folders to not overwrite each other text encoder cache right? @kohya-ss

**Comment by FurkanGozukara** [comment]
> > if we are training with different captions on same dataset, we should use different folders to not overwrite each other text encoder cache right? @kohya-ss
> 
> That's right. Regardless of caching, the only way currently to use different captions for images is to use different folders.

thanks this is important

**Comment by sdbds** [comment]
Hopefully, a multi-level subtitle cache can be implemented, plus it would be nice to be able to set the REPEAT or weights individually.According to the training report of Dalle-3, the ratio of tags to natural language descriptions is best 5% and 95%.
We can set 1:3:6, tags : short descriptions :long descriptions

**Comment by johnr14** [comment]
**And here I was happy to have 3-4 captions in my .allcaps and enabled `shuffle_caption = true` in my dataset.toml !**

So now I just learned that I have to duplicate all datasets x number of caption/tags lines or : 
1. stop training
2. delete caches
3. shuffle .txt file lines `find ./ -type f -name "*.allcaps" -exec sort -R -o {} {} \;`
4. resume training
5. regenerate all chaches
6. hope it works

This is a major problem that captions can't be cached as multiple different strings.  **Can't we implement a quick fix** and just duplicate the cache file. **It's `_flux_te.npz` for text encoder right ?** So just having a few more `_flux_te_0X.npz` would not be worse than duplicating the dataset right ? 

### All in for a new dataformat then !

I find that enabling multiple captions + multiple crop in a single .safetensors is a great idea and this should be pushed.
I can think of having zoomed crops of hands, feets, faces, clothing, accessories ...
Mixing cropping size would be great and each crop could have crop_x_caption1 ...

**Comment by FurkanGozukara** [comment]
@kohya-ss it would be so amazing if you could embed some metadata to decide whether re-cache or not

like when prompt changed or make newer version i dont know

now i am making different folders for every training :D

**Comment by recris** [comment]
Are we planning to support multi-channel alpha masks? I've been using a customized loss function that needs multiple pieces of information (values) for each pixel. It would be nice if the masks could support more than one value per spatial element.
